pkg/terraformer/config.go

```diff
@@ -159,8 +174,31 @@ func (f initializerFunc) Initialize(config *InitializerConfig) error {
 }
 
 // DefaultInitializer is an Initializer that initializes the configuration, variables and state resources
+// based on the given main, variables and tfvars content and on the given InitializerConfig.
+func DefaultInitializer(c client.Client, main, variables string, tfvars []byte) Initializer {
```

Comment:
DefaultInitializer() and StateInitializer() are almost the same besides one function call. WDYT about handing over the state as a *string and decide based on that wether to either create or restore the data?

This would also mean that the [function here](https://github.com/gardener/gardener-extension-provider-aws/blob/0f6900492fc5ac741f00ffa0b680c79e84c0923e/pkg/controller/infrastructure/actuator.go#L69) could have a state parameter  (*runtime.RawExtension)  instead of the unused infrastructure and just unmarshals & hand over if it is not nil.

Reply:
@danielfoehrKn Initially this was the case but we diced to trigger restoration only if operation annotation "Restore" occurs. So the other case is to past state as a *string to the inner reconciliation function and not extract it in it from Infrastructure.
```
func (a *actuator) Restore(...) error {
        terraformState, err := terraformer.UnmarshalRawState(infrastructure.Status.State)
        if err != nil {
		return nil, err
	}
	
	infrastructureStatus, state, err := Reconcile(ctx, a.logger, a.RESTConfig(), a.Client(), a.Decoder(), a.ChartRenderer(), infrastructure, &terraformState.Data)
	...
}
```

Reply:
I'd vote to only use a single method. Perhaps you could pass 
```
if _, err := CreateOrUpdateStateConfigMap(ctx, c, config.Namespace, config.StateName, state); err != nil && !apierrors.IsAlreadyExists(err) {
    return err
}
```
and 
```
if err := CreateStateConfigMap(ctx, c, config.Namespace, config.StateName); err != nil && !apierrors.IsAlreadyExists(err) {
    return err
}
```
As parameters as well.